ci/configs/llvm.sh

```diff
@@ -0,0 +1,15 @@
+CI_DESC="CI job using LLVM-based libraries and tools (clang, libc++, clang-tidy, iwyu) and testing Ninja"
+CI_DIR=build-llvm
+NIX_ARGS=(
+  --arg libcxx true
+)
+CMAKE_ARGS=(
+  -G Ninja
+  -DCMAKE_CXX_COMPILER=clang++
+  -DCMAKE_CXX_STANDARD_LIBRARY=libc++
```

Comment:
I doubt that relying on the undocumented `CMAKE_CXX_STANDARD_LIBRARY` behaves as one might expect in this context. We definitely want:

```suggestion
  -DCMAKE_CXX_COMPILER="clang++;-stdlib=libc++"
```

Reply:
The suggested approach makes `-stdlib=libc++` a permanent part of the compiler invocation, whereas `CXXFLAGS` may be altered by the build system.

Reply:
> If you're saying CMake might randomly alter those flags...

I'm not blaming CMake. Just assuming that `CMakeListss.txt` can potentially process `CMAKE_CXX_FLAGS` in some way. I agree this sounds a bit vague. If it helps, here's a quote form the CMake [docs](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER.html):
> **Note:** Options that are required to make the compiler work correctly can be included as items in a list; they can not be changed.

I'm not opposed to using your suggestion.

Reply:
FWIW, neither of these suggested changes seem to work and both result in:

```
clang++: error: argument unused during compilation: '-stdlib=libc++' [-Werror,-Wunused-command-line-argument]
```

because the stdlib option is a linker option not a compile option. If I add the flag to CMAKE_EXE_LINKER_FLAGS it avoids that error, but then linking fails with lots of undefined symbol errors because libstdc++ headers are included but libc++ is linked.

Setting CMAKE_CXX_STANDARD_LIBRARY seems to be a clean way of choosing the library and avoiding this problem, and is only being set as part of this CI configuration, not part of the project itself. So I think I will just add a comment remarking that this option is experimental and may not work in the future.

There seem to be other ways of choosing the library (https://chatgpt.com/share/685bf4fc-b99c-800a-9342-d1d7458178b5) but this is simpler than any alternative and will probably be documented eventually and if it stops working the CI job can be changed without affecting the project itself.

Reply:
> Looking into the verbose build log, I still can't convince myself that using the `CMAKE_CXX_STANDARD_LIBRARY` variable is a valid approach.

Not sure if you saw the chatgpt summary, but it linked to https://gitlab.kitware.com/cmake/cmake/-/issues/20851 which implements this setting and https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html which says internal variables "are subject to change, and not recommended for use in project code."

This PR is only using the setting in a CI config file that's not part of project code and nothing downstream will use or rely on. So I don't think there is a problem here, but if there is one I could try an alternate approach or just drop commit 403f63a891b1cf0a41e2771ef99261e34ae1f082 since it is not essential

Reply:
Oh nice find. So I guess the option doesn't even do anything here? The other changes in shell.nix do seem to work and Cap'n Proto binaries are built with libc++, but I can confirm ldd shows libmultiprocess binaries using a mix of libc++ and libstdc++ so CMAKE_CXX_STANDARD_LIBRARY is not having the desired effect.

```$ ldd build-llvm/test/mptest 
	linux-vdso.so.1 (0x00007f781796b000)
	libkj-test.so.1.1.0 => /nix/store/q4dhry7vwy2pyybgipnqmnrwnc3x62qn-capnproto-1.1.0/lib/libkj-test.so.1.1.0 (0x00007f781781a000)
	libcapnp-rpc.so.1.1.0 => /nix/store/q4dhry7vwy2pyybgipnqmnrwnc3x62qn-capnproto-1.1.0/lib/libcapnp-rpc.so.1.1.0 (0x00007f7817711000)
	libcapnp.so.1.1.0 => /nix/store/q4dhry7vwy2pyybgipnqmnrwnc3x62qn-capnproto-1.1.0/lib/libcapnp.so.1.1.0 (0x00007f7817652000)
	libkj-async.so.1.1.0 => /nix/store/q4dhry7vwy2pyybgipnqmnrwnc3x62qn-capnproto-1.1.0/lib/libkj-async.so.1.1.0 (0x00007f7817572000)
	libkj.so.1.1.0 => /nix/store/q4dhry7vwy2pyybgipnqmnrwnc3x62qn-capnproto-1.1.0/lib/libkj.so.1.1.0 (0x00007f78174de000)
	libpthread.so.0 => /nix/store/cg9s562sa33k78m63njfn1rw47dp9z0i-glibc-2.40-66/lib/libpthread.so.0 (0x00007f78174d9000)
	libstdc++.so.6 => /nix/store/7c0v0kbrrdc2cqgisi78jdqxn73n3401-gcc-14.2.1.20250322-lib/lib/libstdc++.so.6 (0x00007f7817200000)
	libm.so.6 => /nix/store/cg9s562sa33k78m63njfn1rw47dp9z0i-glibc-2.40-66/lib/libm.so.6 (0x00007f7817117000)
	libgcc_s.so.1 => /nix/store/7c0v0kbrrdc2cqgisi78jdqxn73n3401-gcc-14.2.1.20250322-lib/lib/libgcc_s.so.1 (0x00007f78174ab000)
	libc.so.6 => /nix/store/cg9s562sa33k78m63njfn1rw47dp9z0i-glibc-2.40-66/lib/libc.so.6 (0x00007f7816e00000)
	/nix/store/cg9s562sa33k78m63njfn1rw47dp9z0i-glibc-2.40-66/lib/ld-linux-x86-64.so.2 => /nix/store/cg9s562sa33k78m63njfn1rw47dp9z0i-glibc-2.40-66/lib64/ld-linux-x86-64.so.2 (0x00007f781796d000)
	libc++.so.1 => /nix/store/1gpni0qga8q9jkagxr9m4xzfzanq8avf-libcxx-20.1.5/lib/libc++.so.1 (0x00007f7816ceb000)
	libc++abi.so.1 => /nix/store/1gpni0qga8q9jkagxr9m4xzfzanq8avf-libcxx-20.1.5/lib/libc++abi.so.1 (0x00007f78170cc000)
	libatomic.so.1 => /nix/store/7c0v0kbrrdc2cqgisi78jdqxn73n3401-gcc-14.2.1.20250322-lib/lib/libatomic.so.1 (0x00007f781749d000)
```

I'm very surprised calling a c++ library which is compiled with libc++ from c++ code compiled with libstdc++ could even work, and had assumed if the build was working and tests were passing that would imply the same library was used everywhere. But I guess mixing libraries works because cap'n proto c++ api avoids types like std::string and std::vector, so the two runtime libraries don't actually need to interact with each other.

I'll see if there is some simple way to fix this, otherwise I'll just drop the commit 403f63a891b1cf0a41e2771ef99261e34ae1f082 from the PR. Thanks for catching the problem!

Reply:
re: https://github.com/bitcoin-core/libmultiprocess/pull/184#discussion_r2167002748

> The using of the libstdc++ is obvious.

This should be fixed now. Now it links with only libc++ not libc++ and libstdc++ when I run locally.

re: https://github.com/bitcoin-core/libmultiprocess/pull/184#discussion_r2167002748

> That means, if you set `CMAKE_CXX_FLAGS` (which you should never do), then `CXXFLAGS` may not be used (wich is the reason why you should not set `CMAKE_CXX_FLAGS`).

@purpleKarrot just saw this comment (sorry the thread appears all mixed up). It sound like you're asking to change the two config files to use CXXFLAGS environment variable instead of CMAKE_CXX_FLAGS command line option. I think that makes sense because even though CMAKE_CXX_FLAGS is empty by default, hypothetically if a toolchain or similar file might wanted to influence `CMAKE_CXX_FLAGS` with `CMAKE_CXX_FLAGS_INIT` the command line here would clobber that value while the environment variable would be combined with it.

I guess this logic also extends to CMAKE_CXX_COMPILER and CXX would be recommended instead?

I'll try switching to environment variables in the next push if that makes sense.